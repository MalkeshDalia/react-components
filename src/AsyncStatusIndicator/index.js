import AsyncStatusIndicator, { Props as AsyncStatusIndicatorProps, AsyncStatusIndicatorState } from './AsyncStatusIndicator';

export { AsyncStatusIndicatorProps, AsyncStatusIndicatorState };
export default AsyncStatusIndicator;
