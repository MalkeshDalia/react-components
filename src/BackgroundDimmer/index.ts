import BackgroundDimmer, { BackgroundDimmerProps, Props } from './BackgroundDimmer';
export default BackgroundDimmer;
export { BackgroundDimmerProps, Props };
