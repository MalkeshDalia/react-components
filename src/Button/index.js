export default from './Button';
export StatefulButton from './StatefulButton';
