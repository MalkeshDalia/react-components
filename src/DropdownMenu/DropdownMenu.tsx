import * as React from 'react';
import { props, t, ReactChildren } from '../utils';
import Menu, { optionType } from './Menu';
import FlexView from 'react-flexview';
import Icon from '../Icon/Icon';
import cx from '../utils/classnames';
import { MenuProps } from './Menu';

export namespace DropdownMenuProps {
  export type size =  'small' | 'medium' | 'large';
}

export type DropdownMenuRequiredProps = {
  /** menu button content */
  children?: JSX.Element | string,
  /** renderer for menu items */
  menuRenderer?: (options: MenuProps.Option[]) => JSX.Element,
  /** menu options */
  options: MenuProps.Option[],
  /** className for menu button icon (if children is passed, this is ignored) */
  iconClassName?: string,
  /** called when menu is open */
  onOpen: () => void,
  /** called when menu is closed */
  onClose: () => void,
  /** the height of the menu button */
  size?: DropdownMenuProps.size,
  /** menu button max-height */
  maxHeight?: number,
  className?: string
}

export type DropdownMenuDefaultProps = {
  /** whether the menu is open or not */
  isOpen: boolean,
  /** whether the menu should be closed when clicking outside the dropdown */
  dismissOnClickOut: boolean
}

export type DropdownMenuProps = DropdownMenuRequiredProps & Partial<DropdownMenuDefaultProps>;
type DropdownMenuDefaultedProps = DropdownMenuRequiredProps & DropdownMenuDefaultProps;

export const Props = {
  children: t.maybe(ReactChildren),
  menuRenderer: t.maybe(t.Function),
  options: t.list(optionType),
  iconClassName: t.maybe(t.String), // if children is passed, this is ignored
  isOpen: t.maybe(t.Boolean),
  onOpen: t.Function,
  onClose: t.Function,
  dismissOnClickOut: t.maybe(t.Boolean),
  size: t.maybe(t.enums.of(['small', 'medium', 'large']), ''),
  maxHeight: t.maybe(t.Number),
  className: t.maybe(t.String)
};

/**
 *  A toggleable dropdown menu
 */
@props(Props)
export default class DropdownMenu extends React.PureComponent<DropdownMenuProps> {

  static defaultProps: DropdownMenuDefaultProps = {
    isOpen: false,
    dismissOnClickOut: true
  };

  toggleMenu = () => {
    const { isOpen, onOpen, onClose } = this.props as DropdownMenuDefaultedProps;
    if (isOpen) {
      onClose();
    } else {
      onOpen();
    }
  };

  onMenuClick = () => {
    this.toggleMenu();
  };

  getHeightFromSize = (size?: DropdownMenuProps.size) => {
    switch (size) {
      case 'small': return 250;
      case 'medium': return 400;
      case 'large': return 600;
      default: return null;
    }
  };

  templateToggler = ({ children, iconClassName, isOpen }: {
    children?: JSX.Element | string,
    iconClassName?: string,
    isOpen: boolean
  }) => {
    return (
      children || this.templateIconButton({ iconClassName, isOpen })
    );
  };

  templateIconButton = ({ iconClassName, isOpen }: { iconClassName?: string, isOpen: boolean }) => {
    return (
      <FlexView vAlignContent='center' className={cx('dropdown-menu-icon-container', { isOpen })}>
        <Icon icon={iconClassName} className='dropdown-menu-icon' />
      </FlexView>
    );
  };

  templateMenu = ({ isOpen, options, height, onMenuClick }: {
    isOpen: boolean,
    options: MenuProps.Option[],
    height: number | null,
    onMenuClick: () => void
  }) => {
    return isOpen ? (
      <FlexView onClick={onMenuClick}>
        {this.props.menuRenderer ?
          this.props.menuRenderer(options) : <Menu options={options} style={{ maxHeight: height }} />
        }
      </FlexView>
    ) : null;
  };

  templateOverlay = ({ isOpen, toggleMenu, dismissOnClickOut }: {
    isOpen: boolean,
    toggleMenu: () => void,
    dismissOnClickOut: boolean
  }) => {
    return (
      dismissOnClickOut && isOpen ? <div className='dropdown-menu-overlay' onClick={toggleMenu} /> : null
    );
  };

  render() {
    const {
      iconClassName,
      children,
      options,
      dismissOnClickOut,
      maxHeight,
      className,
      size,
      isOpen
    } = this.props as DropdownMenuDefaultedProps;
    const { toggleMenu, onMenuClick } = this;

    const height = maxHeight || this.getHeightFromSize(size);

    return (
      <FlexView vAlignContent='center' className={cx('dropdown-menu', className)} onClick={toggleMenu}>
        {this.templateOverlay({ isOpen, toggleMenu, dismissOnClickOut })}
        {this.templateToggler({ children, iconClassName, isOpen })}
        {this.templateMenu({ isOpen, options, height, onMenuClick })}
      </FlexView>
    );
  }

}
