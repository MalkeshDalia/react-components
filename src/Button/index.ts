import Button from './Button';
export default Button;
import StatefulButton from './StatefulButton';
export { StatefulButton };
