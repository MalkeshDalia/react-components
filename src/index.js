import Dropdown from './Dropdown';
export { Dropdown };
import LoadingSpinner from './LoadingSpinner';
export { LoadingSpinner };
import MobileDetector from './MobileDetector';
export { MobileDetector };
import Popover from './Popover';
export { Popover };
import Overflow from './Overflow';
export { Overflow };
import TextOverflow from './TextOverflow';
export { TextOverflow };
import ScrollView from './ScrollView/ScrollView';
export { ScrollView };
export { Toaster, TimerToast } from './Toaster';
import Modal from './Modal';
export { Modal };
export { modalWithContext } from './Modal';
import TransitionWrapper from './TransitionWrapper';
export { TransitionWrapper };
import BackgroundDimmer from './BackgroundDimmer';
export { BackgroundDimmer };
import Divider from './Divider';
export { Divider };
import Icon from './Icon';
export { Icon };
export { Panel, PanelHeader, TabbedPanel } from './Panel';
export { Menu, DropdownMenu } from './DropdownMenu';
import Tooltip from './Tooltip';
export { Tooltip };
import Button from './Button';
export { Button };
export { StatefulButton } from './Button';
import MoreOrLess from './MoreOrLess';
export { MoreOrLess };
import ConfirmationInput from './ConfirmationInput';
export { ConfirmationInput };
import Badge from './Badge';
export { Badge };
import NavBar from './NavBar';
export { NavBar };
import Toggle from './Toggle';
export { Toggle };
import Meter from './Meter';
export { Meter };
import CollapsibleSection from './CollapsibleSection';
export { CollapsibleSection };
import BrowserDetector from './BrowserDetector';
export { BrowserDetector };
import TimePicker from './TimePicker';
export { TimePicker };
import FocusableView from './FocusableView';
export { FocusableView };
import ResizeSensor from './ResizeSensor';
export { ResizeSensor };
import Tablo from './Tablo';
export { Tablo };
export { Column as TabloColumn, Cell, Header, Footer } from './Tablo';
import AsyncStatusIndicator from './AsyncStatusIndicator';
export { AsyncStatusIndicator };
import FormattedText from './FormattedText';
export { FormattedText };
import DatePicker from './DatePicker';
export { DatePicker };
import DateField from './DateField';
export { DateField };
import Image from './Image';
export { Image };
