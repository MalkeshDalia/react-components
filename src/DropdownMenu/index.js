import DropdownMenu from './DropdownMenu';
export { DropdownMenu };
import Menu from './Menu';
export { Menu };