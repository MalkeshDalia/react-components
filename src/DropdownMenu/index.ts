import { DropdownMenu } from './DropdownMenu';
import { Menu } from './Menu';

export default DropdownMenu;
export { Menu, DropdownMenu }

