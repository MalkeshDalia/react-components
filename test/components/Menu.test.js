import React from 'react';
import { mount } from 'enzyme';
import { Menu } from '../../src/DropdownMenu';

let consoleError: jest.SpyInstance<{}>;

beforeAll(() => {
  consoleError = jest.spyOn(console, 'error');
});

afterEach(() => {
  expect(consoleError).not.toHaveBeenCalled();
});

const exampleProps = {
  options: [
    { title: 'Preferences', type: 'item' },
    { title: 'Preferences', type: 'item' }
  ]
};

describe('Menu', () => {

  it('renders correctly', () => {
    const component = mount(
      <Menu {...exampleProps} />
    );
    expect(component.html()).toMatchSnapshot();
  });

  it('triggers onClick only on options that are not disabled', () => {
    const onClickEnabled = jest.fn();
    const onClickDisabled = jest.fn();
    const component = mount(
      <Menu options={[
        { title: 'Preferences', type: 'item', onClick: onClickEnabled },
        { title: 'Preferences', type: 'item', disabled: true, onClick: onClickDisabled }
      ]}
      />
    );
    component.find('.menu-item').at(0).simulate('click');
    component.find('.menu-item').at(1).simulate('click');
    expect(onClickEnabled).toBeCalled();
    expect(onClickDisabled).not.toBeCalled();
  });

});
