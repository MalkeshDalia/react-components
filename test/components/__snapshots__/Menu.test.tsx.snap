// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correctly 1`] = `"<div class=\\"menu\\" style=\\"max-height: -20px;\\"><div class=\\"menu-row\\"><div class=\\"react-flex-view align-content-center menu-item\\"><div class=\\"react-flex-view menu-item-title\\">Preferences</div><div class=\\"react-flex-view justify-content-end menu-item-metadata\\"></div></div></div><div class=\\"menu-row\\"><div class=\\"react-flex-view align-content-center menu-item\\"><div class=\\"react-flex-view menu-item-title\\">Preferences</div><div class=\\"react-flex-view justify-content-end menu-item-metadata\\"></div></div></div></div>"`;
